include/tvm/tir/analysis.h

```diff
@@ -86,6 +87,58 @@ TVM_DLL double EstimateTIRFlops(const Stmt& stmt);
  */
 TVM_DLL double EstimateTIRFlops(const IRModule& mod);
 
+/*!
+ * \brief Visitor class to perform use/def analysis, also delete unreferenced lets.
+ * \sa UndefinedVars
+ */
+class VarUseDefAnalyzer : public StmtExprMutator {
```

Comment:
Do we really need to move the mutator to the public header file?

Reply:
In some cases, we might need to visit structures that are neither an `Stmt` nor a `PrimExpr`, with the result type set instead of an array, existing `UndefinedVars` API is not convenient in this case.

Reply:
I agree, the collected variable definitions may be useful in a lot of cases.  Exposing the visitor itself also allows for analyzing multiple expressions, then examining variables that are undefined in any of them.  I'm wondering if there's a tighter API that we'd want to expose, but nothing is coming to mind.

Reply:
Though, could we inherit from `StmtExprVisitor` instead of `StmtExprMutator`?  The only modification done is the deletion of unreferenced lets, but that really feels like something that should be part of the `RemoveNoOp` pass.

Reply:
@Lunderberg I'll try to decouple these functionalities and make it a `StmtExprVisitor`.

Reply:
This is a case where hiding things internally still helps, for advanced use-cases, we can put it as an internal header under src/tir/analysis/var_def_analyzer.h

Reply:
Sounds good, besides the location issue. Should we decouple the Analyzer into 3 functions?
1. VarUseDefAnalyzer
2. DeviceInfoCollector
3. SimplifyingLetStmt

I suppose 2 and 3 are not necessary for the general def-use analysis case but necessary for `SplitHostDevice` pass, I prefer doing the refactor and only keeping `VarUseDefAnalyzer` in the header, and making 2 & 3 parts of `split_host_device.cc`.

Reply:
That works